### PR TITLE
Add comment parser to highlight comment tags

### PR DESCRIPTION
**Pull request by stsewd**
Closes https://github.com/nvim-treesitter/nvim-treesitter/issues/236

### CODE REVIEW

**Comment by vigoux** [comment]
Hey, to be honet with you I am _really_ reticent to add a heavily C-based parser...

Could you tell me the reason of using so much C in your comment parser ?

It is not that I am against this (I will fuzz this parser to be sure everything works great anyway), I am even really happy tat this issue is finally fixed, I just want some information before.

**Comment by stsewd** [comment]
> Hey, to be honet with you I am really reticent to add a heavily C-based parser...

Feels like so much criticism for c-based parsers isn't fair, looks like it's only based on the bad experience of the markdown grammar. There are a lot of parsers that use a custom scanner (because it's needed).

> Could you tell me the reason of using so much C in your comment parser ?

tree-sitter doesn't have a way to fallback nodes that don't match something (they are parsed as an error), in this case the fallback node is `_text`. You can't write this kind of rules in pure js (things that depend on white spaces/indentation or need to parse just some nodes and put everything else under another node). To write them in js means to put everything under a single token (so no separation of symbols like `:`, `(`, `(user)`), or write a really long and complex regex that captures the node and other that matches everything but those nodes (which I think it is maybe impossible to write something like that without the full regex syntax or custom logic). So, yes, I found those limitations while trying to avoid C in the first place, but I wasn't able to find a good result, I can re-write it in js if someone finds a way around those limitations, I can even give it a last try today (maybe just parse known/common tags instead of `{ANYTHING}:`).

> It is not that I am against this (I will fuzz this parser to be sure everything works great anyway), I am even really happy tat this issue is finally fixed, I just want some information before.

I think most of the parsers that can crash nvim are those that serialize data for the external parser (which this one doesn't use, bc it doesn't need to keep a state), plus, this doesn't use any call to malloc and similiar.

**Comment by theHamsta** [comment]
> tree-sitter doesn't have a way to fallback nodes that don't match something

It has. this what declaring `conflicts` in grammars is made for.

Maybe it would be good if you could use standard library function if available for things like `is_whitespace` (c has `iswspace`) and use static linkage. It would be safer to either declare these functions `inline` or `static` or give them unambiguous prefix. GCC will per-default export all symbols from you shared object any dynamic resolving of a symbol might as well resolve to one of your symbols.

**Comment by stsewd** [comment]
> It has. this what declaring conflicts in grammars is made for.

As explained in https://github.com/nvim-treesitter/nvim-treesitter/issues/872#issuecomment-770400369 it doesn't work like you think.

> Maybe it would be good if you could use standard library function if available for things like is_whitespace

Those don't include the EOF mark, I don't see that causing problems.

**Comment by stsewd** [comment]
> It could be problematic to use tree-sitters regex-based system or a pure-C parser in comments where we should expect weird international characters and unicode math symbols. Would be a big plus if this is tested in with this parser.

TS uses `int32_t` for the characters, the grammar only allows ascii uppercase letters for the tag name https://github.com/stsewd/tree-sitter-comment#syntax, it shouldn't be any problem. There is a test for non-ascii characters https://github.com/stsewd/tree-sitter-comment/blob/14fe54e08c890743be677de5703570d311770c94/test/corpus/source.txt#L19

**Comment by stsewd** [comment]
With the latest version I'm able to highlight mentions and issues references. But we could not include them by default if it's too much.

Screenshot from 2021-02-04 00-16-34

**Comment by vigoux** [comment]
It looks great and soi d reasonable, once laŭ disable them by changing highlights.

**Comment by kyazdani42** [comment]
I'm all for it, and looks ready to me. Is it ready to be merged @stsewd ?

**Comment by stsewd** [comment]
@kyazdani42 yeah it's ready! I've been using this branch since I opened this PR, no problems so far! I didn't add the injection to js because looks like comments are already being injected with jsdoc (but I think it's only meant to be injected in docstrings, so someone more familiar with the js ecosystem should edit that file)

https://github.com/nvim-treesitter/nvim-treesitter/blob/5757f8a50d5d26b8c184b3a51713db763cdd9702/queries/javascript/injections.scm#L1-L1

**Comment by kyazdani42** [comment]
@steelsojka not too sure, could we mix up the comment tags parser and the jsdoc parser ?

**Comment by theHamsta** [comment]
@kyazdani42 I would not mix up an externally maintained parser for a specific language with this one that is though generic for all languages.

**Comment by kyazdani42** [comment]
i meant using both parsers for the same range and highlighting twice. I'm not sure how to handle it otherwise.

**Comment by stsewd** [comment]
I was thinking more like use jsdoc for comments that are over a functions/classes and use the comment parser for "simple comments" like inline comments.

**Comment by steelsojka** [comment]
@stsewd Is actually used in all comments for JS since you can have link references in any comment

`// Blah blah blah {@link some/ref/path}`

These two parsers CAN coexist. You can inject multiple languages for the same node, but like highlights, the later match will take priority over previous ones. In the example above, the jsdoc injection will apply it's highlights to that row and then the comment parser would apply it's highlights. If there is no conflicts in highlighting then it would work fine.

**Comment by kyazdani42** [comment]
thanks @steelsojka that was what i was thinking.
@stsewd i guess you could just add the same injection for jsdoc and comment tags and will work just fine. maybe applying jsdoc last would be better.

**Comment by stsewd** [comment]
Having both highlight the same area feels weird, but let me know what you all agree, and I can change it. Also, this can be done in another PR, so this isn't blocked p:

**Comment by steelsojka** [comment]
I'm fine doing this in another pr... And actually when you inject a
language it's already getting highlighted twice the same way. Take for
example a js template literal. It's first highlighted as a string and then
the injected language adds highlights over top of it. As long as the
highlights aren't competing it shouldn't be a problem. The jsdoc highlights
should only really be highlighting tags and not the entire comment
string... Just jsdoc specific syntax.

File: doc/nvim-treesitter.txt
Comment (vigoux) [inline doc/nvim-treesitter.txt:606]: That should be `informational note` IMO.

File: queries/php/injections.scm
Comment (vigoux) [inline queries/php/injections.scm:1]: As you changed that in ocamllex, could umyou also remove the parens here?

File: queries/rust/injections.scm
Comment (vigoux) [inline queries/rust/injections.scm:9]: From what I remember rust also has `block_comment` which should be highlighted too?

**Comment by stsewd** [comment]
Updated with the suggestions from the review and added injections for some new languages!

**Comment by hanh090** [comment]
is it good for merge now?

**Comment by stsewd** [comment]
> is it good for merge now?

yeah, just waiting for an approval :)

File: queries/cpp/injections.scm
Comment (theHamsta) [inline queries/cpp/injections.scm:1]: nope, we can't inject C in C++ proproc macros!

**Review by theHamsta — COMMENTED** [review]
Really nice this feature! Your parser also seems to pass all fuzzing tests.

Could you remove the inherits from the C++ injections (I think there is not self-injection yet)

File: queries/cpp/injections.scm
Comment (stsewd) [reply queries/cpp/injections.scm]: ah, good catch!

**Review by theHamsta — APPROVED** [review]

**Comment by stsewd** [comment]
Looks like all comments have been addressed, thanks everyone for the review!

**Pull request merged**
